File: compiler/packages/babel-plugin-react-compiler/src/ReactiveScopes/PruneUnusedScopes.ts
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {
  ReactiveFunction,
  ReactiveScopeBlock,
  ReactiveStatement,
  ReactiveTerminalStatement,
} from '../HIR/HIR';
import {
  ReactiveFunctionTransform,
  Transformed,
  visitReactiveFunction,
} from './visitors';

// Converts scopes without outputs into regular blocks.
export function pruneUnusedScopes(fn: ReactiveFunction): void {
  visitReactiveFunction(fn, new Transform(), {
    hasReturnStatement: false,
  } as State);
}

type State = {
  hasReturnStatement: boolean;
};

class Transform extends ReactiveFunctionTransform<State> {
  override visitTerminal(stmt: ReactiveTerminalStatement, state: State): void {
    this.traverseTerminal(stmt, state);
    if (stmt.terminal.kind === 'return') {
      state.hasReturnStatement = true;
    }
  }
  override transformScope(
    scopeBlock: ReactiveScopeBlock,
    _state: State,
  ): Transformed<ReactiveStatement> {
    const scopeState: State = {hasReturnStatement: false};
    this.visitScope(scopeBlock, scopeState);
    if (
      !scopeState.hasReturnStatement &&
      scopeBlock.scope.reassignments.size === 0 &&
      (scopeBlock.scope.declarations.size === 0 ||
        /*
         * Can prune scopes where all declarations bubbled up from inner
         * scopes
         */
        !hasOwnDeclaration(scopeBlock))
    ) {
      return {
        kind: 'replace',
        value: {
          kind: 'pruned-scope',
          scope: scopeBlock.scope,
          instructions: scopeBlock.instructions,
        },
      };
    } else {
      return {kind: 'keep'};
    }
  }
}

/*
 * Does the scope block declare any values of its own? This can return
 * false if all the block's declarations are propagated from nested scopes.
 */
function hasOwnDeclaration(block: ReactiveScopeBlock): boolean {
  for (const declaration of block.scope.declarations.values()) {
    if (declaration.scope.id === block.scope.id) {
      return true;
    }
  }
  return false;
}

File: compiler/crates/react_compiler_reactive_scopes/src/prune_unused_scopes.rs
// Copyright (c) Meta Platforms, Inc. and affiliates.
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the root directory of this source tree.

//! PruneUnusedScopes — converts scopes without outputs into regular blocks.
//!
//! Corresponds to `src/ReactiveScopes/PruneUnusedScopes.ts`.

use react_compiler_hir::{
    PrunedReactiveScopeBlock, ReactiveFunction, ReactiveStatement, ReactiveTerminal,
    ReactiveTerminalStatement, ReactiveScopeBlock,
    environment::Environment,
};

use crate::visitors::{
    ReactiveFunctionTransform, Transformed, transform_reactive_function,
};

struct State {
    has_return_statement: bool,
}

/// Converts scopes without outputs into pruned-scopes (regular blocks).
/// TS: `pruneUnusedScopes`
pub fn prune_unused_scopes(func: &mut ReactiveFunction, env: &Environment) {
    let mut transform = Transform { env };
    let mut state = State {
        has_return_statement: false,
    };
    transform_reactive_function(func, &mut transform, &mut state);
}

struct Transform<'a> {
    env: &'a Environment,
}

impl<'a> ReactiveFunctionTransform for Transform<'a> {
    type State = State;

    fn visit_terminal(
        &mut self,
        stmt: &mut ReactiveTerminalStatement,
        state: &mut State,
    ) {
        self.traverse_terminal(stmt, state);
        if matches!(stmt.terminal, ReactiveTerminal::Return { .. }) {
            state.has_return_statement = true;
        }
    }

    fn transform_scope(
        &mut self,
        scope: &mut ReactiveScopeBlock,
        _state: &mut State,
    ) -> Transformed<ReactiveStatement> {
        let mut scope_state = State {
            has_return_statement: false,
        };
        self.visit_scope(scope, &mut scope_state);

        let scope_id = scope.scope;
        let scope_data = &self.env.scopes[scope_id.0 as usize];

        if !scope_state.has_return_statement
            && scope_data.reassignments.is_empty()
            && (scope_data.declarations.is_empty()
                || !has_own_declaration(scope_data, scope_id))
        {
            // Replace with pruned scope
            Transformed::Replace(ReactiveStatement::PrunedScope(
                PrunedReactiveScopeBlock {
                    scope: scope.scope,
                    instructions: std::mem::take(&mut scope.instructions),
                },
            ))
        } else {
            Transformed::Keep
        }
    }
}

/// Does the scope block declare any values of its own?
/// Returns false if all declarations are propagated from nested scopes.
/// TS: `hasOwnDeclaration`
fn has_own_declaration(
    scope_data: &react_compiler_hir::ReactiveScope,
    scope_id: react_compiler_hir::ScopeId,
) -> bool {
    for (_, decl) in &scope_data.declarations {
        if decl.scope == scope_id {
            return true;
        }
    }
    false
}
